src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/TraceEventExtensions.cs

```diff
@@ -65,53 +67,269 @@ public static bool TryGetCounterPayload(this TraceEvent traceEvent, CounterFilte
+                // this happens when the Gauge callback function throws an exception.
+                payload = new CounterEndedPayload(meterName, instrumentName, null, obj.TimeStamp);
+            }
+        }
+
+        private static void HandleCounterRate(TraceEvent traceEvent, CounterFilter filter, string sessionId, out ICounterPayload payload)
         {
-            var metadataDict = new Dictionary<string, string>();
+            payload = null;
+
+            string payloadSessionId = (string)traceEvent.PayloadValue(0);
 
-            ReadOnlySpan<char> metadata = metadataPayload;
+            if (payloadSessionId != sessionId)
+            {
+                return;
+            }
 
-            while (!metadata.IsEmpty)
+            string meterName = (string)traceEvent.PayloadValue(1);
+            //string meterVersion = (string)obj.PayloadValue(2);
+            string instrumentName = (string)traceEvent.PayloadValue(3);
+            string unit = (string)traceEvent.PayloadValue(4);
+            string tags = (string)traceEvent.PayloadValue(5);
+            string rateText = (string)traceEvent.PayloadValue(6);
+
+            if (!filter.IsIncluded(meterName, instrumentName))
             {
-                int commaIndex = metadata.IndexOf(',');
+                return;
+            }
 
-                ReadOnlySpan<char> kvPair;
+            if (double.TryParse(rateText, out double rate))
+            {
+                payload = new RatePayload(meterName, instrumentName, null, unit, tags, rate, filter.IntervalSeconds, traceEvent.TimeStamp);
+            }
+            else
+            {
+                // for observable instruments we assume the lack of data is meaningful and remove it from the UI
+                // this happens when the ObservableCounter callback function throws an exception.
+                payload = new CounterEndedPayload(meterName, instrumentName, null, traceEvent.TimeStamp);
+            }
+        }
+
+        private static void HandleHistogram(TraceEvent obj, CounterFilter filter, string sessionId, out ICounterPayload payload)
+        {
+            payload = null;
+
+            string payloadSessionId = (string)obj.PayloadValue(0);
+            if (payloadSessionId != sessionId)
+            {
+                return;
+            }
+
+            string meterName = (string)obj.PayloadValue(1);
+            //string meterVersion = (string)obj.PayloadValue(2);
+            string instrumentName = (string)obj.PayloadValue(3);
+            string unit = (string)obj.PayloadValue(4);
+            string tags = (string)obj.PayloadValue(5);
+            string quantilesText = (string)obj.PayloadValue(6);
+
+            if (!filter.IsIncluded(meterName, instrumentName))
+            {
+                return;
+            }
 
-                if (commaIndex < 0)
+            if (string.IsNullOrEmpty(quantilesText))
```

Comment:
@noahfalk after histogram values are no longer reported regularly in the app, the quantiles here become empty. is this by design?

Reply:
Yep, the app hasn't disposed the Histogram and also hasn't recorded any data within the measurement interval then by design you get an event where the quantiles string is empty. There is a test case for that behavior here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs#L374

When doing the design initially I thought that sending an explicit event with empty content would let tools know with confidence that the app didn't produce data whereas if an event is absent it is ambiguous what the app did.

Reply:
Should this then return a `CounterEndedPayload` instance?

Reply:
I'll double check this behavior, but I believe it produces more data if more data occurs. I will change our pipeline to produce the event with no quantiles, and we can simply discard them during consumption.

Reply:
Yes, the MetricEventSource events should behave the way you are describing Wiktor. You could choose to model it different ways when you are processing it but looking at Prometheus probably provides a good point of comparison. You could make a Prometheus.NET histogram, record no data, and see what shows up on the metrics endpoint.